sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala

```diff
@@ -163,6 +165,76 @@ case class Expm1(child: Expression) extends UnaryMathExpression(math.expm1, "EXP
 
 case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLOOR")
 
+object Factorial {
+
+  def factorial(n: Int): Long = {
+    if (n < factorials.length) factorials(n) else Long.MaxValue
+  }
+
+  private val factorials: Array[Long] = Array[Long](
+    1,
+    1,
+    2,
+    6,
+    24,
+    120,
+    720,
+    5040,
+    40320,
+    362880,
+    3628800,
+    39916800,
+    479001600,
+    6227020800L,
+    87178291200L,
+    1307674368000L,
+    20922789888000L,
+    355687428096000L,
+    6402373705728000L,
+    121645100408832000L,
+    2432902008176640000L
+  )
+}
+
+case class Factorial(child: Expression)
+  extends UnaryExpression with Serializable with ExpectsInputTypes {
+  override def expectedChildTypes: Seq[DataType] = Seq(IntegerType)
+  override def dataType: DataType = LongType
+  override def foldable: Boolean = child.foldable
+  override def nullable: Boolean = true
```

Comment:
Can you add a comment for this nullable?
E.g. even the `child.nullable=false`, however if the value not in the range of [0-20], it still will be null.